Source/WebGPU/WGSL/Overload.cpp

```diff
@@ -56,11 +56,18 @@ struct ViableOverload {
 
 class OverloadResolver {
 public:
-    OverloadResolver(TypeStore&, const Vector<OverloadCandidate>&, const Vector<Type*>&, unsigned numberOfTypeSubstitutions, unsigned numberOfValueSubstitutions);
+    OverloadResolver(TypeStore&, const Vector<OverloadCandidate>&, const Vector<Type*>& valueArguments, const Vector<Type*>& typeArguments, unsigned numberOfTypeSubstitutions, unsigned numberOfValueSubstitutions);
 
     Type* resolve();
 
 private:
+    struct Substitution {
+        enum class Inferred : bool { No, Yes };
+
+        Type* type { nullptr };
+        Inferred inferred { Inferred::Yes };
+    };
```

Comment:
Conceptually, WGSL builtin types are a finite flat set. Some of those types are spelled with `<` characters in them (e.g. `vec<f32, 4>` but the language doesn't actually support arbitrary substitution. So, I'm curious why you're implementing this infrastructure to handle an arbitrary case, rather than just enumerating all the builtin types, and treating them all as siblings.